Source/Details/ASCollectionViewLayout.m

```diff
@@ -0,0 +1,162 @@
+}
+
+- (instancetype)init
+{
+  return [self initWithLayoutCalculator:[[ASCollectionViewFlowLayoutCalculator alloc] init]];
+}
+
+#pragma mark - ASCollectionViewLayoutCalculatorProviding
+
+- (id<ASCollectionViewLayoutCalculating>)collectionViewLayoutCalculator
+{
+  return _calculator;
+}
+
+- (CGSize)sizeForCalculatingCollectionViewLayout
+{
+  ASDisplayNodeAssertMainThread();
+  return self.collectionView.bounds.size;
+}
+
+#pragma mark - ASCollectionViewLayoutCalculatorResultConsuming
+
+- (void)setLayout:(ASLayout *)layout calculatedBy:(id<ASCollectionViewLayoutCalculating>)calculator forElementMap:(ASElementMap *)map
+{
+  ASDisplayNodeAssertMainThread();
+  ASDisplayNodeAssertTrue(calculator == _calculator);
+  [self invalidateLayout];
+  _collectionViewIsPendingLayoutUpdate = YES;
+  _elementMap = map;
+  [self _applyLayout:layout];
+}
+
+#pragma mark - UICollectionViewLayout overrides
+
+- (void)prepareLayout
+{
+  if (_collectionViewIsPendingLayoutUpdate == NO && _elementMap != nil) {
+    ASLayout *layout = [_calculator layoutThatFits:self.collectionView.bounds.size forElementMap:_elementMap];
+    [self _applyLayout:layout];
+  }
+}
+
+- (void)invalidateLayout
+{
+  [super invalidateLayout];
+  if (_collectionViewIsPendingLayoutUpdate == NO) {
+    _contentSize = CGSizeZero;
+    [_elementToLayoutAttributesMap removeAllObjects];
+  }
+}
+
+- (CGSize)collectionViewContentSize
+{
+  _collectionViewIsPendingLayoutUpdate = NO;
+  return _contentSize;
+}
+
+- (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
+{
+  NSMutableArray *attributesInRect = [NSMutableArray array];
+  NSEnumerator<UICollectionViewLayoutAttributes *> *enumerator = [_elementToLayoutAttributesMap objectEnumerator];
+  UICollectionViewLayoutAttributes *attrs;
+  while ((attrs = [enumerator nextObject])) {
+    if (CGRectIntersectsRect(rect, attrs.frame)) {
+      [attributesInRect addObject:attrs];
+    }
+  }
+  return attributesInRect;
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+  ASCollectionElement *element = [_elementMap elementForItemAtIndexPath:indexPath];
+  return [_elementToLayoutAttributesMap objectForKey:element];
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath
+{
+  ASCollectionElement *element = [_elementMap supplementaryElementOfKind:elementKind atIndexPath:indexPath];
+  return [_elementToLayoutAttributesMap objectForKey:element];
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForDecorationViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath
+{
+  return [self layoutAttributesForSupplementaryViewOfKind:elementKind atIndexPath:indexPath];
+}
+
+- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds
+{
+  if (!CGSizeEqualToSize(self.collectionView.bounds.size, newBounds.size)) {
```

Comment:
I'm not sure, but should this use sizeForCalculatingCollectionViewLayout in some way?  Maybe the API should be sizeFor...LayoutWithBounds: to make it more functional based on the bounds, and then we could compare the current size for layout with the size that would be used once the new bounds is applied.